### DISCUSSION TITLE
Is there any difference between the shell-operator and policy engines?

### DISCUSSION THREAD

**Discussion by wyfeng001** [flant/shell-operator#585] [General]
As the title, Is there any big difference between this project and the Kubernetes dynamic control project in CNCF? e.g. Kyverno [https://kyverno.io/docs/introduction/](https://kyverno.io/docs/introduction/).  kubewarden [https://www.kubewarden.io/ ](https://www.kubewarden.io/ )  OPA Gatekeeper [https://open-policy-agent.github.io/gatekeeper/website/docs/](https://open-policy-agent.github.io/gatekeeper/website/docs/) and so on

**Comment by nabokihms** [comment]
Hello, @wyfeng001.

Shell-operator is more generic than policy engines. There are some notable use cases for it:
1. Operator (like the one written with the controller-runtime in go). It can react to events in the cluster, like creating or deleting objects in Kubernetes.
2. Metrics exporter. It can subscribe to kubernetes objects and expose metrics.
3. Validating/Mutation Admission Controller like policy engines you've mentioned.
